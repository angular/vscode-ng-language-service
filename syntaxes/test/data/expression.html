<!-- Data types -->
{{ variable }}
{{ 'string' }}
{{ "string" }}
{{ 1 }}
{{ true }}
{{ false }}
{{ null }}
{{ this }}
{{ undefined }}
{{ [variable, variable] }}
{{ ['string', 'string'] }}
{{ ["string", "string"] }}
{{ [1, 2] }}
{{ [true, false, null, this, undefined] }}
{{ {object: variable} }}
{{ {object: 'string'} }}
{{ {object: "string"} }}
{{ {object: 1} }}
{{ {object: true} }}
{{ {object: false} }}
{{ {object: null} }}
{{ {object: this} }}
{{ {object: undefined} }}
{{ {object: [variable, variable]} }}
{{ {object: ['string', 'string']} }}
{{ {object: ["string", "string"]} }}
{{ {object: [1, 2]} }}
{{ {object: [true, false, null, this, undefined]} }}

<!-- Function call and property read -->
{{ call() }}
{{ call(); }}
{{ call().object }}
{{ call().object; }}
{{ call()?.conditional }}
{{ call()?.conditional; }}
{{ call()!.conditional }}
{{ call()!.conditional; }}

<!-- Nested function call -->
{{ object.call() }}
{{ object.call(); }}
{{ object.call().object }}
{{ object.call().object; }}
{{ object?.call()?.conditional }}
{{ object?.call()?.conditional; }}
{{ object!.call()!.conditional }}
{{ object!.call()!.conditional; }}

<!-- Function parameter test -->
{{ call(withParams) }}
{{ call('withParams') }}
{{ call("withParams") }}
{{ call("withParams") }}
{{ call(false) }}
{{ call(!!true) }}
{{ call(["array", 123]) }}
{{ call(1) }}
{{ call({ test: 'object' }) }}
{{ call({ test: 123 }) }}
{{ call({ test: [123, 321, { value: ['string', false, true] }] }) }}

<!-- Ternary expression -->
{{ condition ? true : false }}
{{ condition ? true : false; }}
{{ condition() ? call(1 + 2 + 3) : call()  }}
{{ condition() ? call(1 + 2 + 3) : call();  }}
{{ condition()?.object ? call()!.test() : false }}
{{ condition ? ['123'] : { test: 'a' } }}
{{ condition ? 'test' : "test" }}
{{ condition ? [function(), variable] : {} }}

<!-- Microsyntax test -->
{{ (let param of params?.anotherParam!.param().param | async) as p; let i = index; let first = first }}
{{ let param of params | async as p; let i = index; let first = first; }}
{{ (let param of params | async) as p; index as i; first as first; }}
{{ value | pipe:'with-param' }}
{{ { test: 'value'} | json }}
{{ value && value || value && !!value }}
