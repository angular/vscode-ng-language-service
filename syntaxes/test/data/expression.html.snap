><!-- Interpolation test -->
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div>{{ true ? call(1 + 2 + 3) : 'test'  }}</div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div>{{ testing.call(1 + 2 + 3, another?.one()).test }}</div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>
><td mat-cell
#^^^^^^^^^^^^^ expression.ng
>    *ngFor="let param of params | async as p; last as last; first as first">
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>    {{ param.costCode?.name(test, {test: 'test'} 'test', 123) }}
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
></td>
#^^^^^^ expression.ng
>
><!-- Property binding test -->
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div [ ngStyle ]="{
#^^^^^^^^^^^^^^^^^^^^ expression.ng
>  'max-width.px': i * 2 + 5
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>}"></div>
#^^^^^^^^^^ expression.ng
><div [ngClass]="{
#^^^^^^^^^^^^^^^^^^ expression.ng
>    'class-one': test.test(),
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>    'class-two',
#^^^^^^^^^^^^^^^^^ expression.ng
>    'class-three'
#^^^^^^^^^^^^^^^^^^ expression.ng
>}"></div>>
#^^^^^^^^^^^ expression.ng
>
><div [@animation.trigger]="val"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><img [attr.aria-label]="val" />
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div [my-property]="val"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div [my_property]="val"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div [myProperty$]="val"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div [%invalidProperty]="val"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div [invalidProperty)="val"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>
><!-- Event binding test -->
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><button (click)="onClick($event)"></button>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><input (ngModelChange)="onModelChange($event)" />
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div (@animation.done)="onAnimationDone($event)"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div (someEvent)="
#^^^^^^^^^^^^^^^^^^^ expression.ng
>    if (isCondition) {
#^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>        methodIfTrue($event);
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>    } else {
#^^^^^^^^^^^^^ expression.ng
>        methodIfFalse($event);
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>    }
#^^^^^^ expression.ng
>"></div>
#^^^^^^^^^ expression.ng
><div ( extraSpacing )="onExtraSpacing($event)"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component (myEvent)="onMyEvent($event)"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component (my-event)="onMyEvent($event)"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component (my_event)="onMyEvent($event)"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component (myEvent$)="onMyEvent($event)"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component (%invalidEvent)="onMyEvent($event)"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component (invalidEvent]="onMyEvent($event)"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>
><!-- Two-way binding test -->
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><button [(click)]="clickProp"></button>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div [( extraSpacing )]="extraSpacing"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div [(@animation.done)]="animation"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component [(my-prop)]="myProp"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component [(my_prop)]="myProp"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component [($my_prop)]="myProp"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component [(%invalid)]="invalid"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component ([invalid)]="invalid"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>
><!-- Event binding test -->
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><button *ngModel="title"></button>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><div *ngFor="let book of books"></div>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component *custom-if="true != false"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component *custom_if="true != false"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component *custom_$if="true != false"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
><my-component *%invalid="expr"></my-component>
#^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expression.ng
>